### DISCUSSION TITLE
为什么 DNS 使用 UDP 协议 · Why's THE Design? · /whys-the-design-dns-udp-tcp

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#574] [Announcements]
https://draveness.me/whys-the-design-dns-udp-tcp

**Comment by ttttmr** [comment]
学习了，感谢博主

**Comment by silver-brother** [comment]
学习了  大佬，.me的域名你在哪里注册的啊？阿里云好像没有这种域名可以注册

**Comment by draveness** [comment]
> 学习了 大佬，.me的域名你在哪里注册的啊？阿里云好像没有这种域名可以注册

Godaddy 上，你查一下吧

**Comment by ArvonHuang** [comment]
楼主你好，我想了解一下，linux是如何管理内存的，谢谢

**Comment by draveness** [comment]
> 楼主你好，我想了解一下，linux是如何管理内存的，谢谢

好问题，我记一下

**Comment by zuyan9** [comment]
还有一个可能的原因，UDP 比较快吧，打开个新网站多等几个（与 DNS Server 之间的）RTT 还是很烦的

**Comment by draveness** [comment]
> @duguzuyang
> 还有一个可能的原因，UDP 比较快吧，打开个新网站多等几个（与 DNS Server 之间的）RTT 还是很烦的

文中也提到过『多次通信以及信息传输带来的时间成本在 DNS 查询较小时是无法被忽视的』，还有

> 需要注意的是，我们在这里计算结果的前提是 DNS 解析器只需要与一个命名服务器或者权威服务器进行通信就可以获得 DNS 响应，但是在实际场景中，DNS 解析器可能会递归地与多个命名服务器进行通信，这也加倍地放大了 TCP 协议在额外开销上的劣势。

这些原因会放大 TCP 协议在耗时和数据方面的开销

**Comment by wujunze** [comment]
也有 DNS Server  用 TCP 的

**Comment by draveness** [comment]
> 也有 DNS Server 用 TCP 的

这文中不是提到了么...... DNS 协议现在可以用 UDP、TDP、TLS 和 HTTP

**Comment by wujunze** [comment]
> > 也有 DNS Server 用 TCP 的
> 
> 这文中不是提到了么...... DNS 协议现在可以用 UDP、TDP、TLS 和 HTTP

对  哈哈

**Comment by linxiaowu66** [comment]
大佬~~学习了

**Comment by ray1888** [comment]
说一个小补充，内核恐慌podcast也有一集专门提到dns相关的知识，但是更多是向安全性方面的讨论（dns攻击etc....）有兴趣也可以听一下

**Comment by draveness** [comment]
> 说一个小补充，内核恐慌podcast也有一集专门提到dns相关的知识，但是更多是向安全性方面的讨论（dns攻击etc....）有兴趣也可以听一下

点赞，我之前也是内核恐慌的听众

**Comment by anjiadoo** [comment]
发现宝藏，哈哈，等大佬的go出书了我一定会补交学会的

**Comment by draveness** [comment]
> 发现宝藏，哈哈，等大佬的go出书了我一定会补交学会的

可以的

**Comment by tranch** [comment]
「区域传输」这个翻译读起来好别扭，总是联想到地理的区域概念…

**Comment by draveness** [comment]
> 「区域传输」这个翻译读起来好别扭，总是联想到地理的区域概念…

你有啥建议么，没找到合理的引用

**Comment by icbd** [comment]
> @draveness
> > 「区域传输」这个翻译读起来好别扭，总是联想到地理的区域概念…
> 
> 你有啥建议么，没找到合理的引用

"跨区同步" 怎么样?

**Comment by TarryHoo** [comment]
"当客户端接收到一个被阶段的 DNS 响应时" 被截断

---

*2020-09-23 UPDATES：已修复*

**Comment by AlanMaxxx** [comment]
有一个问题是：为什么现在的DNS数据包越来越大呢？根据DNS协议的数据包构造，DNS只是对一个域名进行查找IP地址，意思是域名的名字会越来越长吗？如果是这样，好像也不见的吧？域名有大小。

**Comment by AlanMaxxx** [comment]
我看了附加链接“When do DNS queries use TCP instead of UDP?”，这里面讲得为什么需要TCP是主要因为是DNSSEC或IPV6地址导致传输数据大于UDP报文最大长度，而UDP又是不可靠传输协议，那么为了避免分片UDP数据包导致丢失数据包，所以才使用TCP来传输，而不是你所说的MTU的大小限制，MTU限制主要是链路层的帧大小，对TCP同样限制。

**Comment by draveness** [comment]
> 有一个问题是：为什么现在的DNS数据包越来越大呢？根据DNS协议的数据包构造，DNS只是对一个域名进行查找IP地址，意思是域名的名字会越来越长吗？如果是这样，好像也不见的吧？域名有大小。

你下面的提问其实也回答了这个问题

> 我看了附加链接“When do DNS queries use TCP instead of UDP?”，这里面讲得为什么需要TCP是主要因为是DNSSEC或IPV6地址导致传输数据大于UDP报文最大长度，而UDP又是不可靠传输协议，那么为了避免分片UDP数据包导致丢失数据包，所以才使用TCP来传输，而不是你所说的MTU的大小限制，MTU限制主要是链路层的帧大小，对TCP同样限制。

我重新看了一下这个问题，在 TCP RFC7766 的附录中，出现了不太一样的原因 https://tools.ietf.org/html/rfc7766#appendix-A

>    IP fragmentation is less of a problem for TCP than it is for UDP.
   TCP stacks generally implement Path MTU Discovery so they can avoid
   IP fragmentation of TCP segments.  UDP, on the other hand, does not
   provide reassembly; this means datagrams that exceed the path MTU
   size must experience fragmentation [RFC5405].  Middleboxes are known
   to block IP fragments, leading to timeouts and forcing client
   implementations to "hunt" for EDNS0 reply size values supported by
   the network path.  Additionally, fragmentation may lead to cache
   poisoning [fragmentation-considered-poisonous].

这里还是会采用更可靠的来源做出结论，原链接从这里删掉了

**Comment by StrayLittlePunk** [comment]
何对使用 TCP 协议的 DNS 进行一些优化，减少一些额外开销？
维持一个最大连接时间,在这个时间段发送DNS请求无需额外握手, 打包多个DNS请求发送
请问大佬这样做是生产环境实践吗, 还有更好实践吗? 第二个问题DNS协议还能用什么协议传输呢?

**Comment by songshitong** [comment]
大佬，DNS-over-QUIC 可以解决tcp和udp的问题吗
https://adguard.com/zh_cn/blog/dns-over-quic.html

**Comment by real-jiakai** [comment]
无论是选择 UDP 还是 TCP，最核心的矛盾就在于需要传输的数据包大小，如果数据包小到一定程度，UDP 协议绝对最佳的选择，但是当数据包逐渐增大直到突破 512 字节以及 MTU 1500 字节的限制时，我们也只能选择使用更可靠的 TCP 协议来传输 DNS 查询和相应。:smile:

**Comment by runxii** [comment]
博主写得好棒啊，很清晰明了，排版也好舒服！！想问问展示TCP&UDP PACKET SIZE的图也是博主自己做的吗？是用的什么方式呢，也想学习一下^^

**Reply by runxii** [reply]
不好意思！刚刚看到画图的链接了🙏🏻🙏🏻

**Comment by Elward-lv** [comment]
为什么mongoDb的那篇文章会被无线重定向0-0
